File: frontend-desktop/windows/analytics.py
"""
Advanced Analytics Window Module
Displays comprehensive analytics dashboard with health scores, statistics, outliers, and rankings.
"""

from PyQt5.QtWidgets import (
    QWidget, QVBoxLayout, QHBoxLayout, QLabel, QTableWidget, QTableWidgetItem,
    QScrollArea, QFrame
)
from PyQt5.QtCore import Qt
from PyQt5.QtGui import QFont, QColor
import matplotlib.pyplot as plt
from matplotlib.backends.backend_qt5agg import FigureCanvasQTAgg as FigureCanvas
from matplotlib.figure import Figure
import io


class AnalyticsWindow(QWidget):
    """
    Advanced Analytics Dashboard showing comprehensive equipment analysis.
    Displays health scores, statistics, visualizations, and risk assessment.
    """
    def __init__(self, dataset):
        """
        Initialize analytics window.
        
        Args:
            dataset: Dictionary containing all analytics data
        """
        super().__init__()
        self.dataset = dataset
        self.init_ui()

    def init_ui(self):
        """Initialize and setup the analytics UI"""
        self.setWindowTitle("Advanced Analytics Dashboard")
        self.setGeometry(100, 100, 1200, 800)
        
        layout = QVBoxLayout()
        
        # Title
        title = QLabel("📊 Advanced Analytics Dashboard")
        title_font = QFont()
        title_font.setPointSize(16)
        title_font.setBold(True)
        title.setFont(title_font)
        layout.addWidget(title)
        
        # Summary Cards
        layout.addLayout(self.create_summary_cards())
        
        # Statistics Table
        layout.addWidget(self.create_statistics_table())
        
        # Charts
        layout.addLayout(self.create_charts())
        
        # Efficiency Ranking
        layout.addWidget(self.create_ranking_table())
        
        # Outliers
        if self.dataset.get('outliers'):
            layout.addWidget(self.create_outliers_table())
        
        # Add scroll area
        scroll = QScrollArea()
        scroll.setWidgetResizable(True)
        container = QWidget()
        container.setLayout(layout)
        scroll.setWidget(container)
        
        main_layout = QVBoxLayout()
        main_layout.addWidget(scroll)
        self.setLayout(main_layout)

    def create_summary_cards(self):
        """Create summary metric cards"""
        layout = QHBoxLayout()
        
        # Health Score Card
        health_frame = self.create_card(
            "⭐ Average Health Score",
            f"{self.dataset.get('avg_health_score', 0)}/100",
            "#4CAF50"
        )
        layout.addWidget(health_frame)
        
        # Risk Summary Card
        risk = self.dataset.get('risk_summary', {})
        risk_text = f"High: {risk.get('high_risk', 0)}\nMed: {risk.get('medium_risk', 0)}\nLow: {risk.get('low_risk', 0)}"
        risk_frame = self.create_card(
            "🚨 Risk Summary",
            risk_text,
            "#2196F3"
        )
        layout.addWidget(risk_frame)
        
        # Outliers Card
        outlier_frame = self.create_card(
            "⚠️ Outliers",
            str(self.dataset.get('outlier_count', 0)),
            "#FF9800"
        )
        layout.addWidget(outlier_frame)
        
        return layout

    def create_card(self, title, value, color):
        """Create a metric card"""
        frame = QFrame()
        frame.setStyleSheet(f"""
            QFrame {{
                border-left: 4px solid {color};
                background-color: #f5f5f5;
                border-radius: 4px;
                padding: 15px;
            }}
        """)
        
        layout = QVBoxLayout()
        
        title_label = QLabel(title)
        title_font = QFont()
        title_font.setBold(True)
        title_label.setFont(title_font)
        layout.addWidget(title_label)
        
        value_label = QLabel(value)
        value_font = QFont()
        value_font.setPointSize(14)
        value_font.setBold(True)
        value_label.setFont(value_font)
        value_label.setStyleSheet(f"color: {color};")
        layout.addWidget(value_label)
        
        frame.setLayout(layout)
        return frame

    def create_statistics_table(self):
        """Create statistics summary table"""
        label = QLabel("📈 Statistical Analysis")
        font = QFont()
        font.setBold(True)
        label.setFont(font)
        
        table = QTableWidget()
        table.setColumnCount(6)
        table.setHorizontalHeaderLabels(["Parameter", "Min", "Max", "Median", "Std Dev", "Mean"])
        table.setMaximumHeight(150)
        
        stats = self.dataset.get('statistics', {})
        row = 0
        
        for param_name in ['flowrate', 'pressure', 'temperature']:
            param_data = stats.get(param_name, {})
            table.insertRow(row)
            
            table.setItem(row, 0, QTableWidgetItem(param_name.capitalize()))
            table.setItem(row, 1, QTableWidgetItem(str(param_data.get('min', 'N/A'))))
            table.setItem(row, 2, QTableWidgetItem(str(param_data.get('max', 'N/A'))))
            table.setItem(row, 3, QTableWidgetItem(str(param_data.get('median', 'N/A'))))
            table.setItem(row, 4, QTableWidgetItem(str(param_data.get('std', 'N/A'))))
            table.setItem(row, 5, QTableWidgetItem(str(param_data.get('mean', 'N/A'))))
            
            row += 1
        
        table.resizeColumnsToContents()
        
        layout = QVBoxLayout()
        layout.addWidget(label)
        layout.addWidget(table)
        
        widget = QWidget()
        widget.setLayout(layout)
        return widget

    def create_charts(self):
        """Create visualization charts"""
        layout = QHBoxLayout()
        
        # Pressure vs Temperature Scatter
        if self.dataset.get('equipment_data'):
            scatter_canvas = self.create_scatter_chart()
            layout.addWidget(scatter_canvas)
            
            # Health Score Distribution
            health_canvas = self.create_health_chart()
            layout.addWidget(health_canvas)
        
        return layout

    def create_scatter_chart(self):
        """Create pressure vs temperature scatter plot"""
        fig = Figure(figsize=(5, 4), dpi=100)
        ax = fig.add_subplot(111)
        
        equipment_data = self.dataset.get('equipment_data', [])
        pressures = [eq['pressure'] for eq in equipment_data]
        temperatures = [eq['temperature'] for eq in equipment_data]
        health_scores = [eq['health_score'] for eq in equipment_data]
        
        scatter = ax.scatter(pressures, temperatures, c=health_scores, cmap='RdYlGn', s=50, alpha=0.6)
        ax.set_xlabel("Pressure")
        ax.set_ylabel("Temperature")
        ax.set_title("Pressure vs Temperature")
        fig.colorbar(scatter, ax=ax, label="Health Score")
        
        canvas = FigureCanvas(fig)
        return canvas

    def create_health_chart(self):
        """Create health score distribution bar chart"""
        fig = Figure(figsize=(5, 4), dpi=100)
        ax = fig.add_subplot(111)
        
        equipment_data = self.dataset.get('equipment_data', [])[:10]
        names = [eq['name'][:10] for eq in equipment_data]
        scores = [eq['health_score'] for eq in equipment_data]
        colors = ['green' if s >= 85 else 'orange' if s >= 70 else 'red' for s in scores]
        
        ax.barh(names, scores, color=colors)
        ax.set_xlabel("Health Score")
        ax.set_title("Equipment Health Scores")
        ax.set_xlim(0, 100)
        
        canvas = FigureCanvas(fig)
        return canvas

    def create_ranking_table(self):
        """Create efficiency ranking table"""
        label = QLabel("🏆 Equipment Efficiency Ranking")
        font = QFont()
        font.setBold(True)
        label.setFont(font)
        
        table = QTableWidget()
        table.setColumnCount(5)
        table.setHorizontalHeaderLabels(["Rank", "Equipment Name", "Type", "Health Score", "Status"])
        table.setMaximumHeight(300)
        
        ranking = self.dataset.get('efficiency_ranking', [])
        
        for row, item in enumerate(ranking[:10]):
            table.insertRow(row)
            
            table.setItem(row, 0, QTableWidgetItem(str(item['rank'])))
            table.setItem(row, 1, QTableWidgetItem(item['equipment_name']))
            table.setItem(row, 2, QTableWidgetItem(item['type']))
            
            score_item = QTableWidgetItem(str(item['health_score']))
            score_item.setForeground(QColor("#2196F3"))
            table.setItem(row, 3, score_item)
            
            status_item = QTableWidgetItem(item['status'])
            if item['status'] == 'Excellent':
                status_item.setForeground(QColor("#4CAF50"))
            elif item['status'] == 'Good':
                status_item.setForeground(QColor("#2196F3"))
            elif item['status'] == 'Fair':
                status_item.setForeground(QColor("#FF9800"))
            else:
                status_item.setForeground(QColor("#F44336"))
            table.setItem(row, 4, status_item)
        
        table.resizeColumnsToContents()
        
        layout = QVBoxLayout()
        layout.addWidget(label)
        layout.addWidget(table)
        
        widget = QWidget()
        widget.setLayout(layout)
        return widget

    def create_outliers_table(self):
        """Create outliers table"""
        label = QLabel("⚠️ Detected Outliers")
        font = QFont()
        font.setBold(True)
        label.setFont(font)
        
        table = QTableWidget()
        table.setColumnCount(7)
        table.setHorizontalHeaderLabels(["Equipment", "Type", "Flowrate", "Pressure", "Temperature", "Health Score", "Risk"])
        table.setMaximumHeight(250)
        
        outliers = self.dataset.get('outliers', [])
        
        for row, outlier in enumerate(outliers):
            table.insertRow(row)
            
            table.setItem(row, 0, QTableWidgetItem(outlier['equipment_name']))
            table.setItem(row, 1, QTableWidgetItem(outlier['type']))
            
            params = outlier['parameters']
            table.setItem(row, 2, QTableWidgetItem(str(params['flowrate'])))
            table.setItem(row, 3, QTableWidgetItem(str(params['pressure'])))
            table.setItem(row, 4, QTableWidgetItem(str(params['temperature'])))
            
            score_item = QTableWidgetItem(str(outlier['health_score']))
            table.setItem(row, 5, score_item)
            
            risk_item = QTableWidgetItem(outlier['risk'])
            if outlier['risk'] == 'HIGH':
                risk_item.setForeground(QColor("#F44336"))
            elif outlier['risk'] == 'MEDIUM':
                risk_item.setForeground(QColor("#FF9800"))
            else:
                risk_item.setForeground(QColor("#4CAF50"))
            table.setItem(row, 6, risk_item)
        
        table.resizeColumnsToContents()
        
        layout = QVBoxLayout()
        layout.addWidget(label)
        layout.addWidget(table)
        
        widget = QWidget()
        widget.setLayout(layout)
        return widget

File: frontend-web/src/components/Analytics.js
import React from "react";
import { ScatterChart, Scatter, BarChart, Bar, XAxis, YAxis, CartesianGrid, Tooltip, Legend, ResponsiveContainer, PieChart, Pie, Cell } from "recharts";
import "./Analytics.css";

function Analytics({ summary, equipment_data }) {
  if (!summary || !equipment_data) return null;

  const { statistics, equipment_data: equipmentList, efficiency_ranking, avg_health_score, risk_summary, outliers } = summary;

  // ============ SCATTER PLOT DATA: Pressure vs Temperature ============
  const scatterData = equipmentList.map((eq) => ({
    pressure: eq.pressure,
    temperature: eq.temperature,
    name: eq.name,
    health_score: eq.health_score,
  }));

  // ============ HISTOGRAM DATA: Flowrate Distribution ============
  const flowrateBins = [[], [], [], [], []];
  const minFlow = statistics.flowrate.min;
  const maxFlow = statistics.flowrate.max;
  const binSize = (maxFlow - minFlow) / 5;

  equipmentList.forEach((eq) => {
    const binIndex = Math.min(4, Math.floor((eq.flowrate - minFlow) / binSize));
    flowrateBins[binIndex].push(eq);
  });

  const histogramData = flowrateBins.map((bin, i) => ({
    range: `${(minFlow + i * binSize).toFixed(1)}-${(minFlow + (i + 1) * binSize).toFixed(1)}`,
    count: bin.length,
  }));

  // ============ HEATMAP DATA: Equipment Type × Parameter Matrix ============
  const typeStats = {};
  equipmentList.forEach((eq) => {
    if (!typeStats[eq.type]) {
      typeStats[eq.type] = {
        type: eq.type,
        flowrate: [],
        pressure: [],
        temperature: [],
      };
    }
    typeStats[eq.type].flowrate.push(eq.flowrate);
    typeStats[eq.type].pressure.push(eq.pressure);
    typeStats[eq.type].temperature.push(eq.temperature);
  });

  const heatmapData = Object.values(typeStats).map((item) => ({
    type: item.type,
    avg_flowrate: (item.flowrate.reduce((a, b) => a + b, 0) / item.flowrate.length).toFixed(1),
    avg_pressure: (item.pressure.reduce((a, b) => a + b, 0) / item.pressure.length).toFixed(1),
    avg_temperature: (item.temperature.reduce((a, b) => a + b, 0) / item.temperature.length).toFixed(1),
  }));

  // ============ HEALTH SCORE DISTRIBUTION ============
  const healthScores = equipmentList.map((eq) => ({
    name: eq.name,
    score: eq.health_score,
  }));

  // ============ RISK DISTRIBUTION (PIE CHART) ============
  const riskData = [
    { name: "High Risk", value: risk_summary.high_risk, fill: "#ff4444" },
    { name: "Medium Risk", value: risk_summary.medium_risk, fill: "#ffaa00" },
    { name: "Low Risk", value: risk_summary.low_risk, fill: "#44aa44" },
  ];

  return (
    <div className="analytics-container">
      <h2>📊 Advanced Analytics Dashboard</h2>

      {/* ============ SUMMARY CARDS ============ */}
      <div className="summary-cards">
        <div className="card health-card">
          <h3>⭐ Average Health Score</h3>
          <div className="score">{avg_health_score}</div>
          <p>out of 100</p>
        </div>
        <div className="card risk-card">
          <h3>🚨 Risk Summary</h3>
          <div className="risk-items">
            <div className="risk-high">High: {risk_summary.high_risk}</div>
            <div className="risk-medium">Medium: {risk_summary.medium_risk}</div>
            <div className="risk-low">Low: {risk_summary.low_risk}</div>
          </div>
        </div>
        <div className="card outlier-card">
          <h3>⚠️ Outliers Detected</h3>
          <div className="score">{outliers.length}</div>
          <p>equipment items</p>
        </div>
      </div>

      {/* ============ STATISTICAL SUMMARY TABLE ============ */}
      <div className="stats-table-section">
        <h3>📈 Statistical Analysis</h3>
        <table className="stats-table">
          <thead>
            <tr>
              <th>Parameter</th>
              <th>Min</th>
              <th>Max</th>
              <th>Median</th>
              <th>Std Dev</th>
              <th>Mean</th>
            </tr>
          </thead>
          <tbody>
            <tr>
              <td><strong>Flowrate</strong></td>
              <td>{statistics.flowrate.min}</td>
              <td>{statistics.flowrate.max}</td>
              <td>{statistics.flowrate.median}</td>
              <td>{statistics.flowrate.std}</td>
              <td>{statistics.flowrate.mean}</td>
            </tr>
            <tr>
              <td><strong>Pressure</strong></td>
              <td>{statistics.pressure.min}</td>
              <td>{statistics.pressure.max}</td>
              <td>{statistics.pressure.median}</td>
              <td>{statistics.pressure.std}</td>
              <td>{statistics.pressure.mean}</td>
            </tr>
            <tr>
              <td><strong>Temperature</strong></td>
              <td>{statistics.temperature.min}</td>
              <td>{statistics.temperature.max}</td>
              <td>{statistics.temperature.median}</td>
              <td>{statistics.temperature.std}</td>
              <td>{statistics.temperature.mean}</td>
            </tr>
          </tbody>
        </table>
      </div>

      {/* ============ VISUALIZATIONS ============ */}
      <div className="visualizations">
        
        {/* Scatter Plot: Pressure vs Temperature */}
        <div className="chart-section">
          <h3>📍 Pressure vs Temperature Correlation</h3>
          <ResponsiveContainer width="100%" height={300}>
            <ScatterChart margin={{ top: 20, right: 20, bottom: 20, left: 20 }}>
              <CartesianGrid strokeDasharray="3 3" />
              <XAxis type="number" dataKey="pressure" name="Pressure" />
              <YAxis type="number" dataKey="temperature" name="Temperature" />
              <Tooltip cursor={{ strokeDasharray: "3 3" }} />
              <Legend />
              <Scatter
                name="Equipment"
                data={scatterData}
                fill="#8884d8"
                shape="circle"
              />
            </ScatterChart>
          </ResponsiveContainer>
        </div>

        {/* Histogram: Flowrate Distribution */}
        <div className="chart-section">
          <h3>📊 Flowrate Distribution</h3>
          <ResponsiveContainer width="100%" height={300}>
            <BarChart data={histogramData}>
              <CartesianGrid strokeDasharray="3 3" />
              <XAxis dataKey="range" />
              <YAxis />
              <Tooltip />
              <Bar dataKey="count" fill="#82ca9d" />
            </BarChart>
          </ResponsiveContainer>
        </div>

        {/* Heatmap Table: Equipment Type × Parameters */}
        <div className="chart-section">
          <h3>🔥 Equipment Type Performance Matrix</h3>
          <table className="heatmap-table">
            <thead>
              <tr>
                <th>Equipment Type</th>
                <th>Avg Flowrate</th>
                <th>Avg Pressure</th>
                <th>Avg Temperature</th>
              </tr>
            </thead>
            <tbody>
              {heatmapData.map((item, idx) => (
                <tr key={idx}>
                  <td><strong>{item.type}</strong></td>
                  <td>{item.avg_flowrate}</td>
                  <td>{item.avg_pressure}</td>
                  <td>{item.avg_temperature}</td>
                </tr>
              ))}
            </tbody>
          </table>
        </div>

        {/* Health Score Bar Chart */}
        <div className="chart-section">
          <h3>💪 Equipment Health Scores</h3>
          <ResponsiveContainer width="100%" height={400}>
            <BarChart data={healthScores} layout="vertical">
              <CartesianGrid strokeDasharray="3 3" />
              <XAxis type="number" domain={[0, 100]} />
              <YAxis dataKey="name" type="category" width={120} />
              <Tooltip />
              <Bar dataKey="score" fill="#8884d8" />
            </BarChart>
          </ResponsiveContainer>
        </div>

        {/* Risk Distribution Pie Chart */}
        <div className="chart-section">
          <h3>⭕ Risk Distribution</h3>
          <ResponsiveContainer width="100%" height={300}>
            <PieChart>
              <Pie
                data={riskData}
                cx="50%"
                cy="50%"
                labelLine={true}
                label={({ name, value }) => `${name}: ${value}`}
                outerRadius={80}
                fill="#8884d8"
                dataKey="value"
              >
                {riskData.map((entry, index) => (
                  <Cell key={`cell-${index}`} fill={entry.fill} />
                ))}
              </Pie>
              <Tooltip />
            </PieChart>
          </ResponsiveContainer>
        </div>
      </div>

      {/* ============ EFFICIENCY RANKING ============ */}
      <div className="ranking-section">
        <h3>🏆 Equipment Efficiency Ranking</h3>
        <div className="ranking-list">
          {efficiency_ranking.slice(0, 10).map((item, idx) => (
            <div key={idx} className={`ranking-item rank-${item.status.toLowerCase()}`}>
              <span className="rank-badge">{item.rank}</span>
              <div className="rank-details">
                <strong>{item.equipment_name}</strong> ({item.type})
              </div>
              <div className="rank-score">
                <span className="health-score">{item.health_score}</span>
                <span className="status">{item.status}</span>
              </div>
            </div>
          ))}
        </div>
      </div>

      {/* ============ OUTLIERS ============ */}
      {outliers.length > 0 && (
        <div className="outliers-section">
          <h3>⚠️ Detected Outliers</h3>
          <div className="outlier-list">
            {outliers.map((item, idx) => (
              <div key={idx} className={`outlier-item risk-${item.risk.toLowerCase()}`}>
                <div className="outlier-name">
                  <strong>{item.equipment_name}</strong> ({item.type})
                </div>
                <div className="outlier-params">
                  Flowrate: {item.parameters.flowrate} | Pressure: {item.parameters.pressure} | Temperature: {item.parameters.temperature}
                </div>
                <div className="outlier-score">
                  Health Score: <strong>{item.health_score}</strong> | Risk: <span className={`risk-${item.risk.toLowerCase()}`}>{item.risk}</span>
                </div>
              </div>
            ))}
          </div>
        </div>
      )}
    </div>
  );
}

export default Analytics;
